src/audio/igo_nr/igo_nr.c

```diff
@@ -275,124 +277,96 @@ static inline int32_t set_capture_func(struct comp_dev *dev)
 	return 0;
 }
 
-static struct comp_dev *igo_nr_new(const struct comp_driver *drv,
-				   const struct comp_ipc_config *config, const void *spec)
+static int igo_nr_init(struct processing_module *mod)
 {
-	const struct ipc_config_process *ipc_igo_nr = spec;
-	struct comp_dev *dev = NULL;
-	struct comp_data *cd = NULL;
-	size_t bs = ipc_igo_nr->size;
+	struct module_data *md = &mod->priv;
+	struct comp_dev *dev = mod->dev;
+	struct module_config *cfg = &md->cfg;
+	struct comp_data *cd;
+	size_t bs = cfg->size;
 	int32_t ret;
 
-	comp_cl_info(&comp_igo_nr, "igo_nr_new()");
+	comp_info(dev, "igo_nr_init()");
 
 	/* Check first that configuration blob size is sane */
 	if (bs > SOF_IGO_NR_MAX_SIZE) {
-		comp_cl_err(&comp_igo_nr, "igo_nr_new() error: configuration blob size = %u > %d",
-			    bs, SOF_IGO_NR_MAX_SIZE);
-		return NULL;
+		comp_err(dev, "igo_nr_init() error: configuration blob size = %u > %d",
+			 bs, SOF_IGO_NR_MAX_SIZE);
+		return -EINVAL;
 	}
 
-	dev = comp_alloc(drv, sizeof(*dev));
-	if (!dev)
-		return NULL;
-
-	dev->ipc_config = *config;
-
 	cd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*cd));
 	if (!cd)
-		goto fail;
+		return -ENOMEM;
 
+	md->private = cd;
 	ret = IgoLibGetInfo(&cd->igo_lib_info);
 	if (ret != IGO_RET_OK) {
-		comp_cl_err(&comp_igo_nr, "igo_nr_new(): IgoLibGetInfo() Failed.");
+		comp_err(dev, "igo_nr_init(): IgoLibGetInfo() Failed.");
+		ret = -EINVAL;
 		goto cd_fail;
 	}
 
 	cd->p_handle = rballoc(0, SOF_MEM_CAPS_RAM, cd->igo_lib_info.handle_size);
 	if (!cd->p_handle) {
-		comp_cl_err(&comp_igo_nr, "igo_nr_new(): igo_handle memory rballoc error");
+		comp_err(dev, "igo_nr_init(): igo_handle memory rballoc error for size %d",
+			 cd->igo_lib_info.handle_size);
+		ret = -ENOMEM;
 		goto cd_fail;
 	}
 
-	comp_set_drvdata(dev, cd);
-
 	/* Handler for configuration data */
 	cd->model_handler = comp_data_blob_handler_new(dev);
 	if (!cd->model_handler) {
-		comp_cl_err(&comp_igo_nr, "igo_nr_new(): comp_data_blob_handler_new() failed.");
+		comp_err(dev, "igo_nr_init(): comp_data_blob_handler_new() failed.");
 		goto cd_fail;
```

Comment:
do we need to set an error code here too?

Reply:
Yep